draft-ounsworth-cfrg-kem-combiners.mkd

```diff
@@ -292,43 +292,70 @@ where:
+encoding ciphertext where length is a priori unknown.
 
-~~~ BEGIN EDNOTE ~~~
+## k_i construction {#sec-k_i-construction}
 
-Each `k_i` MUST be constant in length, therefore it MAY be the concatenation of the secret share `ss_i` and ciphertext `ct_i` only if they are guaranteed to be constant length:
+Following the guidance of Giacon et al. [GHP18], we wish for a KEM combiner that is CCA-secure as long as at least one of the ingredient KEMs is. In order to protect against chosen ciphertext attacks, it is necessary to include both the shared secret `ss_i` and its corresponding ciphertext `ct_i`.
+If both the secret share `ss_i` and the ciphertext `ct_i` are constant length, then `k_i` MAY be constructed concatenating the two values:
 
 ~~~
-k_i = ss_i || ct_i
+k_i = ct_i || ss_i
 ~~~
 
-For all other cases, it is REQUIRED to concatenate and hash them first:
+If `ss_i` or `ct_i` are not guaranteed to have constant length, it is REQUIRED to append the `rlen` encoded length when concatenating, prior to inclusion in the overall construction described in {{tab-kemCombiner}}:
 
 ~~~
-k_i = H(ss_i || ct_i)
+k_i = ct_i || rlen(ct_i) || ss_i || rlen(ss_i)
 ~~~
 
-Including the ciphertext guarantees that the combined kem is IND-CCA secure as long as one of the ingredient KEMs is, as stated by [GHP18].
+Any protocols making use of this construction MUST either right-encode the length of all inputs `ss_i` and `ct_i`, or justify that any inputs will always be fixed length.
+In the case of a PSK the associated ciphertext is the empty string.
 
-Any protocols making use of this construction MUST either hash all inputs `ss_i || ct_i`, or justify that any un-hashed inputs will always be fixed length.
+Including the ciphertext guarantees that the combined kem is IND-CCA2 secure as long as one of the ingredient KEMs is, as stated by [GHP18].
 
-~~~ END EDNOTE ~~~
+The ciphertext precedes the secret share, as memory-constrained devices can write `c_i` into the hash state and no further caching is required when streaming.
 
 ## Protocol binding
-The `fixedInfo` string is a fixed-length string containing some context-specific information.
-The intention is to prevent cross-context attacks by making this key derivation unique to its protocol context.
+The `fixedInfo` parameter is a fixed-format string containing some context-specific information.
+This serves to prevent cross-context attacks by making this key derivation unique to its protocol context.
+
+The `fixedInfo` string MUST have a definite structure depending on the protocol where all parts strictly defined by the protocol specification.
+
+~~~
+fixedInfo = fixedInfo || s
+~~~
+
+Each fixed-length input string `f` MAY be directly used as input:
 
-The `fixedInfo` string MUST have a definite structure depending on the protocol where all parts are fixed length. This prevents a variable length structure from creating collisions between two different instances.
```

Comment:
I think requiring fixed length as MUST can be too restritive. Imagine the case where specifiers used for the characterization of an instance have no length bounds per se.

Reply:
This sentence has been in fact removed, I also agree we can allow variable-length, as long as they are suffixed with `rlen`